core/iwasm/common/wasm_shared_memory.c

```diff
@@ -87,6 +93,37 @@ search_module(WASMModuleCommon *module)
     return NULL;
 }
 
+static void
+create_list_of_waiter_addresses(void *key, void *value, void *user_data)
+{
+    AtomicWaitAddressArgs *data = (AtomicWaitAddressArgs *)user_data;
+    data->addr[(data->index++) % WAIT_MAP_SIZE] = key;
+}
+
+void
+notify_stale_threads_on_exception(WASMModuleInstanceCommon *module_inst)
+{
+    AtomicWaitAddressArgs *args = NULL;
```

Comment:
I'd suggest allocating `AtomicWaitAddressArgs` on the stack, but dynamically allocate `addr` field (currently it's fixed value - 32), so you don't have to guess the size of the array in advance.

Reply:
done

Reply:
I see you still allocate it on a heap.